marlowe-actus/src/Language/Marlowe/ACTUS/QCGenerator.hs

```diff
@@ -0,0 +1,180 @@
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# OPTIONS_GHC -fno-warn-incomplete-uni-patterns -fno-warn-name-shadowing -fno-warn-unused-do-bind #-}
+module Language.Marlowe.ACTUS.QCGenerator where
+
+import           Language.Marlowe.ACTUS.Definitions.ContractTerms
+import           Test.QuickCheck
+import           Data.Time
+import           Data.Time.Clock.POSIX(posixSecondsToUTCTime)
+
+epochToDay :: Integer -> Day
+epochToDay = utctDay . posixSecondsToUTCTime . fromIntegral
+
+largeamount :: Gen Double
+largeamount = choose (0.0, 10000000.0)
+
+smallamount :: Gen Double
+smallamount = choose (0.0, 1000.0)
+
+percentage :: Gen Double
+percentage = choose (0.0, 100.0)
+
+scalingFactor :: Gen Double
+scalingFactor = choose (0.0, 100.0)
+
+mightbe :: Gen a -> Gen (Maybe a)
+mightbe original = oneof [ Just <$> original, return Nothing ]
+
+zeroOr :: Gen Double -> Gen Double
+zeroOr original = oneof [return 0.0, original]
+
+oneOr :: Gen Double -> Gen Double
+oneOr original = oneof [return 1.0, original]
+
+date :: Gen Day
+date = epochToDay <$> choose (0, 1000)
+
+datecycle :: Gen Cycle
+datecycle = Cycle <$> choose (0, 100) <*> elements [P_D, P_W, P_M, P_Q, P_H, P_Y] <*> elements [ShortStub, LongStub]
+
+contractTermsGen :: Gen ContractTerms
+contractTermsGen = do 
+    contracttype <- elements [PAM]
+    sd <- date
+    ied <- date
+
+    interest <- mightbe percentage
+    notional <- mightbe largeamount
+    pdied <- zeroOr smallamount
+
+    feebasis <- elements [FEB_A, FEB_N]
+    feerate <- zeroOr percentage
+
+    penaltytype <- elements [PYTP_A , PYTP_N , PYTP_I , PYTP_O]
+    penaltyrate <- zeroOr percentage
+    cpyrt <- zeroOr percentage
+
+    includeendday <- elements [True, False]
+    eomc <- elements [EOMC_EOM, EOMC_SD]
+    bdc <- elements [BDC_NULL, BDC_SCF, BDC_SCMF, BDC_CSF, BDC_CSMF, BDC_SCP, BDC_SCMP, BDC_CSP, BDC_CSMP]
+    dcc <- elements [DCC_A_AISDA, DCC_A_360, DCC_A_365, DCC_E30_360ISDA, DCC_E30_360, DCC_B_252]
+    pref <- elements [PREF_Y, PREF_N]
+    cntrl <- elements [CR_BUY, CR_SEL]
+
+    scef <- elements [SE_000, SE_0N0, SE_00M, SE_0NM, SE_I00, SE_IN0, SE_I0M, SE_INM]
+    scixsd <- oneOr scalingFactor
+    scied <- oneOr scalingFactor
+
+    rrsp <- zeroOr percentage
+    rrmlt <- oneOr scalingFactor
+    rrpf <- zeroOr percentage
+    rrpc <- oneOr percentage
+    rrlc <- oneOr percentage
+    rrlf <- zeroOr percentage
+
+    nextPrincipalRedemption <- mightbe largeamount
+    purchaseDate <- mightbe date
+    maturityDate <- mightbe date
+    terminationDate <- mightbe date
```

Comment:
Another way of doing it is generating a list of dates, ordering them, and assign them in the right implicit order.
So for example, if `purchase < maturity < termination`, then we can generate a list of 3 dates, sort them, and then assign the first to `purchase`, the second to `maturity` and the third to `termination`

Reply:
@palas there is order but I'd like to compare errors with original actus spec with negative tests as those parts are more likely to differ.
It is important because juding from my manual checks we have differences in ways we handle events (which is not specified in the spec). Namely, 

- date collisions: what happens if termination date = maturity date, I add both while java version overwrites those
- dates precedence: what if analysis date is before maturity date but after start date (there unintended side-effects like accruing interest gets incremented), while normally it should be before start date
- utility - if first interest payment happens before initial exchange - what the amount should be: zero, error, negative value?


Although, I gotta agree that if negative cases take 99% of the sample it ain't good either.
 I'll have to test obtain this statistic and rebalance generators if needed.

Reply:
I've created a story: https://jira.iohk.io/browse/SCP-1582.

Basically the change is easy if required - I'll only have to satisfy: IED < any ANX (anchor) < MD

Reply:
Ah, since you mention it: We probably won't get any collisions if the generator just chooses random dates (because the probability is `(1/number_of_possible_dates)^3`, which I am sure is a number ridiculously small. That is another argument in favour of using increments and biasing them to get 0 with a reasonably high probability. Or just restricting the dates to be one out of a finite set of dates. But having a finite set of dates doesn't sound very reassuring to me either.